### PR TITLE
docs: Add type system docs and add details to data source docs

### PR DESCRIPTION
**Pull request by felixwang9817**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: docs/reference/type-system.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Type System
+
+## Motivation
+
+Feast uses an internal type system to provide guarantees on training and serving data.
+Feast currently supports eight primitive types - `INT32`, `INT64`, `FLOAT32`, `FLOAT64`, `STRING`, `BYTES`, `BOOL`, and `UNIX_TIMESTAMP` - and the corresponding array types.
```
Comment (achals) [inline docs/reference/type-system.md:6]: What about `NULL`? And are any of these types nullable, or must they be pure values?

File: docs/reference/type-system.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+The type system is controlled by [`Values.proto`](https://github.com/feast-dev/feast/blob/master/protos/feast/types/Value.proto) in protobuf and by [`types.py`](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/types.py) in Python.
+Type conversion logic can be found in [`type_map.py`](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/type_map.py).
+
+## Examples
```
Comment (achals) [inline docs/reference/type-system.md:10]: Maybe add a section on getting historical features, and another on getting online features?

Comment (felixwang9817) [reply docs/reference/type-system.md:11]: updated

**Review by achals — APPROVED** [review]
/lgtm

**Pull request merged**